### PR TITLE
New plugin: E3DC

### PR DESCRIPTION
**Pull request by SebastianRinger**
nymea-plugins-modbus pull request checklist:

- [x] Make sure the pull request's title is of format "Plugin name: Add support for xyz" or "New plugin: Plugin name"

- [x] Did you test the changes on hardware, if not (e.g. absence of required hardware), please mention a person to confirm it has been tested.

- [x] Did you update the plugin's README.md accordingly?

- [ ] Did you update translations (`cd builddir && make lupdate`)?

### CODE REVIEW

**Comment by mzanetti** [comment]
Hi, thanks for the pull request. Here's some first feedback:

* There is a e3dc-registers.json but it seems incomplete and not used. I would suggest to make use of the modbus tool which generates the modbus tcp connection class based on the register.json, including reconnecting, initial register probing and such. However, seeing that you manually already wrote such a connection class, not sure if you want to switch to that. It would make future changes easier though. In any case, if you decide against that, perhaps delete the unused registers.json altogether.
* There are some stateType maps. This isn't wrong, and is something we've been doing ourselves previously but nowadays you could also just use `thing->setStateValue("connected", ...)` instead and drop those maps (ParamTypes do not support that yet, but will eventually. Just as info, not required to change.
* While the thing classes define states like totalEnergyConsumed, totalEnergyProduced and such, I don't see them actually being set in the code. The plugin seems to only set currentPower. This will cause the energy charts in the app to not show any values. A plugin really should fill in all the state types it defines.
* Coding style: nymea follows the Qt coding style. While we are not 100% strict on it, I think some rough style guide should be followed. Most notably class names should be CamelCased and normally don't have any special characters like _ or so. Also some missing spaces, like `}else{` jump into the eye quickly. Would be nice if you could align those bits a bit.
* The copyright headers are either missing or outdated (year 2021).
* There are classes like Ed3cBattery and such, which have signals for currentPowerChanged but never emit those signals in the .cpp. After looking for a bit I found that the modbus connection class sets the value *and* emits the signal. This is something that makes the code very hard to follow and is considered bad practice in the Qt world. While signals are public (so you can connect to them from the outside) and so the compiler does not complain about this, signals really should only ever be emitted from within their own class. There may be some exceptions to that rule, but in this case I don't see the value for that. Better to just move the emit currentPowerChanged into the setCurrentPower() method. This way one can be sure it will never be forgotten when calling setCurrentPower():
* That said, I am not sure if you really need those classes altogether, given the only thing they do is to store the currentPower value. This is also done by the already existing `Thing`... So in fact, you may just change the connection->updateCurrentPower() method to take the `Thing *` as parameter and directly set it to that. Would make the plugin a lot smaller while achieving exactly the same.
* I can see some potential crashers, especially on shutdown or when the user removes a thing it is probably a 10% chance the plugin will crash, but I'll spare that for the next round. Depending on what you do with the above feedback, those lines of code might not exist any more in the next version.

File: e3dc/integrationplugine3dc.json [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+                            "displayName": "IP address",
+                            "type": "QString",
+                            "defaultValue": "127.0.0.1"
+                        },
+                        {
+                            "id": "1d1768c1-421c-4486-9843-22e4646668b7",
+                            "name":"port",
+                            "displayName": "Port",
+                            "type": "int",
+                            "defaultValue": 502
```
Comment (mzanetti) [inline e3dc/integrationplugine3dc.json:31]: is the port configurable on the device? If yes, ok, if now, this can go away.

File: e3dc/integrationplugine3dc.json [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+                            "displayName": "MAC address",
+                            "type": "QString",
+                            "defaultValue": "00:00:00:00:00:00"
+                        },
+                        {
+                            "id": "ac61f55c-4bc4-4815-9a3e-8e01ad2da739",
+                            "name":"slaveId",
+                            "displayName": "Slave ID",
+                            "type": "int",
+                            "defaultValue": 1
```
Comment (mzanetti) [inline e3dc/integrationplugine3dc.json:45]: same as with port: is the slave address configurable on the device?

File: e3dc/integrationplugine3dc.json [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+                {
+                    "name": "e3dcSmartmeter",
+                    "displayName": "E3dcSmartmeter",
+                    "id": "9ea8830b-3c3f-4ed0-9012-9e7bcd1c25ee",
+                    "createMethods": ["Auto"],
+                    "interfaces": ["energymeter","connectable"],
+                    "paramTypes": [
+                    {
+                        "id": "94904c50-f1fc-41ac-82e0-3bf42334c48c",
+                        "name":"modelId",
```
Comment (mzanetti) [inline e3dc/integrationplugine3dc.json:70]: I don't see the modelId actually being used in the code. Is this supposed to be informational to the user?

**Review by mzanetti — COMMENTED** [review]
Hi, thanks for the contribution.

**Comment by MennerM** [comment]
Hi, 

thanks for the detailed feedback. 
We will take the comments and work on it. The plugin was still developed based on nymea core 1.4.0 and therefore still need to be adapted to the current version.

**Comment by t-mon** [comment]
Any update on this? The integration would be very useful, since you already work on it?

**Comment by mzanetti** [comment]
Closing this as it's really nowhere near complete and it doesn't look like anyone will be working on this any more.
